uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php

```diff
@@ -839,4 +825,181 @@ public static function getTotalTOILBalanceChangeForContact($contactID, $absenceT
+      FROM {$balanceChangeTable} bc
+      LEFT JOIN {$toilRequestTable} toil_request
+            ON bc.source_type = %1 AND bc.source_id = toil_request.id
+      LEFT JOIN {$leaveRequestTable} toil_leave_request
+            ON toil_request.leave_request_id = toil_leave_request.id
+      LEFT JOIN {$periodEntitlementTable} period_entitlement
+            ON bc.source_type = %2 AND bc.source_id = period_entitlement.id
+      LEFT JOIN {$absencePeriodTable} absence_period
+            ON period_entitlement.period_id = absence_period.id
+      WHERE bc.expiry_date IS NOT NULL AND
+            (bc.expiry_date BETWEEN %3 AND %4) AND
+            bc.expiry_date >= %5 AND  
+            bc.expired_balance_change_id IS NULL AND
+            (toil_leave_request.contact_id = %6 OR period_entitlement.contact_id = %6) AND
+            (toil_leave_request.type_id = %7 OR period_entitlement.type_id = %7)    
+      ORDER BY bc.expiry_date ASC, bc.id ASC";
+
+    $params = [
+      1 => [self::SOURCE_TOIL_REQUEST, 'String'],
+      2 => [self::SOURCE_ENTITLEMENT, 'String'],
+      3 => [$absencePeriod->start_date, 'String'],
+      4 => [$absencePeriod->end_date, 'String'],
+      5 => [$leaveRequestFromDate->format('Y-m-d'), 'String'],
+      6 => [$leaveRequest->contact_id, 'Integer'],
+      7 => [$leaveRequest->type_id, 'Integer']
+    ];
+
+    $result = CRM_Core_DAO::executeQuery($query, $params);
+
+    $expiredBalanceChanges = [];
+    while ($result->fetch()) {
+      $expiredBalanceChanges[] = [
+        'id' => $result->id,
+        'type_id' => $result->type_id,
+        'amount' => (float) $result->amount,
+        'start_date' => new DateTime($result->start_date),
+        'expiry_date' => new DateTime($result->expiry_date),
+        'source_type' => $result->source_type,
+        'source_id' => $result->source_id,
+        'contact_id'  => $leaveRequest->contact_id,
+        'absence_type_id' => $leaveRequest->type_id,
+      ];
+    }
+
+    return $expiredBalanceChanges;
+  }
+
+
+  /**
+   * This method checks each LeaveBalanceChanges against the leave dates
+   * taken within the LeaveBalanceChange start and expiry date for same contact and absence type
+   * and deducts the appropriate amount from the remaining amount.
+   *
+   * For expired LeaveBalanceChanges, the record
+   * is simply updated while for  LeaveBalanceChanges just about to be expired,
+   * an expiry record is created with the expired_balance_change_id pointing to the LeaveBalanceChange
+   * to be expired.
+   *
+   * @param array $balanceChanges
+   *   An array of  Leave Balance changes
+   * @param array $datesOverlappingBalanceChanges
+   *   An Array of Leave Request dates overlapping the Leave Balance changes
+   *
+   * @return int
+   *   Number of records created/updated
+   */
+  private static function calculateBalanceChange($balanceChanges, $datesOverlappingBalanceChanges) {
```

Comment:
The name `calculateBalanceChange` is a bit too generic. I think something like `calculateExpiryAmount` would probably be better for this method